### ISSUE TITLE
Strange behaviour when trying to use TT_/TTI_ instructions / use Welford's LLK

### ISSUE THREAD

**Issue by bbradelTT** [tenstorrent/tt-metal#26979]
There are some strange behaviours when trying to use TT_/TTI_ instructions. 

1. Seemingly small changes in what is in an if / else statement have large effects. The easiest to repro is that moving code into an if/else changes amount of cycles taken.

Repro steps:

Run the following test (note that repro steps use the file name test_ln1.py and that the test seems to fail with bad pcc) using rmillerTT/26077_welford_layernorm_infra on tt-metal and vsureshTT/Welfords_LLK on tt-llk with TT_LAYERNORM_WELFORD=1 with changes listed lower down.
```
import pytest
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc

@pytest.mark.parametrize("h", [32])
@pytest.mark.parametrize("w", [64])
@pytest.mark.parametrize("dtype", [torch.bfloat16])
def test_layer_norm_with_weight_bias_and_residual_input(device, h, w, dtype):
    torch.manual_seed(0)

    torch_input_tensor = torch.rand((h, w), dtype=dtype)
    torch_residual_input_tensor = torch.rand((h, w), dtype=dtype)
    torch_weight = torch.rand((w,), dtype=dtype)
    torch_bias = torch.rand((w,), dtype=dtype)
    torch_output_tensor = torch.nn.functional.layer_norm(
        torch_input_tensor + torch_residual_input_tensor, normalized_shape=[w], weight=torch_weight, bias=torch_bias
    )

    input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
    residual_input_tensor = ttnn.from_torch(torch_residual_input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
    weight = ttnn.from_torch(torch_weight, layout=ttnn.TILE_LAYOUT, device=device)
    bias = ttnn.from_torch(torch_bias, layout=ttnn.TILE_LAYOUT, device=device)

    output_tensor = ttnn.layer_norm(input_tensor, residual_input_tensor=residual_input_tensor, weight=weight, bias=bias)
    output_tensor = ttnn.to_layout(output_tensor, ttnn.ROW_MAJOR_LAYOUT)
    output_tensor = ttnn.from_device(output_tensor)
    output_tensor = ttnn.to_torch(output_tensor)

    assert_with_pcc(torch_output_tensor, output_tensor, 0.9997)
```

Something like the following in tt-metal directory:
```
git fetch origin
git checkout rmillerTT/26077_welford_layernorm_infra
git fetch origin
git submodule sync
git submodule update --init --recursive
cd tt_metal/third_party/tt_llk
git fetch origin
git checkout
vsureshTT/Welfords_LLK
cd -
./build_metal.sh -p
```
Replace tt_metal/third_party/tt_llk/tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_welfords.h with the code in https://github.com/tenstorrent/tt-llk/pull/588#issuecomment-3193007206 (the old code is slow, the new code is faster).
```
python -m tracy -r -m pytest test_ln1.py
```
should result in DEVICE FW DURATION [ns] of around 12593

In that h file, in Welfords_Main() change
```
    if (N == 0) {
        Welfords_Load_Data<0,0>();
        INCR_CLOCK();
        WELFORDS_LOOP_ITERATION(N, endN)
    } else {
        Welfords_Load_Initial_Data();
        INCR_CLOCK();
        WELFORDS_LOOP_ITERATION(N, endN)
    }
```
to
```
    if (N == 0) {
        Welfords_Load_Data<0,0>();
    } else {
        Welfords_Load_Initial_Data();
    }
    WELFORDS_LOOP_ITERATION(N, endN)
```
Then re-running the test
```
python -m tracy -r -m pytest test_ln1.py
```
should result in DEVICE FW DURATION [ns] of around 77705

Tested on an N150 with the commits on those branches from Friday afternoon.

2. Also, with the first (faster) code, there seem to be other effects on the order of hundreds of cycles. E.g.
```
    constexpr uint32_t offset1 = (I*32) + (4*J);
    constexpr uint32_t offset2 = offset1 + 1;
    constexpr uint32_t offset3 = offset2 + 1;
    constexpr uint32_t offset4 = offset3 + 1;
            /*row1*/TTI_SFPLOAD(p_sfpu::LREG0, 0, ADDR_MOD_3, offset1);
            /*row2*/TTI_SFPLOAD(p_sfpu::LREG1, 0, ADDR_MOD_3, offset2);
            /*row3*/TTI_SFPLOAD(p_sfpu::LREG2, 0, ADDR_MOD_3, offset3);
            /*row4*/TTI_SFPLOAD(p_sfpu::LREG3, 0, ADDR_MOD_3, offset4);
            /*transposes raw mixed data to logical rows*/
            TTI_SFPTRANSP(0, 0, 0, 0);
            /*past_mean = dst1*/TTI_SFPLOAD(p_sfpu::LREG4, 0, ADDR_MOD_3, 64);
            /*past_var = dst2*/TTI_SFPLOAD(p_sfpu::LREG5, 0, ADDR_MOD_3, 128);
            //wiping LREG 6 and 7 since they may be filled with garbage data
    return;
```
seems to be around  5 cycles while
```
    constexpr uint32_t offset1 = (I*32) + (4*J);
    constexpr uint32_t offset2 = offset1 + 1;
    constexpr uint32_t offset3 = offset2 + 1;
    constexpr uint32_t offset4 = offset3 + 1;
            /*row1*/TTI_SFPLOAD(p_sfpu::LREG0, 0, ADDR_MOD_3, offset1);
            /*row2*/TTI_SFPLOAD(p_sfpu::LREG1, 0, ADDR_MOD_3, offset2);
            /*row3*/TTI_SFPLOAD(p_sfpu::LREG2, 0, ADDR_MOD_3, offset3);
            /*row4*/TTI_SFPLOAD(p_sfpu::LREG3, 0, ADDR_MOD_3, offset4);
            /*transposes raw mixed data to logical rows*/
            TTI_SFPTRANSP(0, 0, 0, 0);
            /*past_mean = dst1*/TTI_SFPLOAD(p_sfpu::LREG4, 0, ADDR_MOD_3, 64);
            /*past_var = dst2*/TTI_SFPLOAD(p_sfpu::LREG5, 0, ADDR_MOD_3, 128);
            //wiping LREG 6 and 7 since they may be filled with garbage data
            TTI_SFPLOADI(p_sfpu::LREG6, 0, 0);
    return;
```
seems to be well over 300 cycles. This seems to be the first time it's called, and that delay may shift around.

Unfortunately this latter effect requires the rest of the llk code to not be called, and the total time is faster than reading in the tile from DRAM and therefore it's not as easily seen in the timing.

One possible approach might be to take the code in https://github.com/tenstorrent/tt-llk/pull/588#issuecomment-3194030929 and compare the call to Welford's in the first loop iteration to the call in the second loop iteration.

I don't know if there could be instruction cache effects, some issues with capturing timestamps, or something else is going on.

**Comment by amahmudTT** [comment]
Had some high level talk with tensix team, the branch at the top will relay all its branch prediction penalty to the tensix engine, as it has to depend on the trisc core to issue its instructions, but the penalty we are observing is still very high. We would need to simulate to understand the final issue.

**Comment by amahmudTT** [comment]
@bbradelTT @vsureshTT There is a bit too much indirection and changes described in the issue. Here are the things i need for simulation. Please provide me with these.

1. Use of minimum size input that manifests the bug. 
2. Two branches, with nearly identical code. Only the code causing the issue should differ. There should not be any DPRINT, or one test be in llk another in metal or anything. Just identical branches with minimal change. Please provide these so i can simulate and talk with tensix people.

**Comment by amahmudTT** [comment]
Had a talk with @vsureshTT , I think the current implementation he has is working properly. The intermediate code described in this issue has slowdowns. I will track that though.

**Comment by bbradelTT** [comment]
Issue is still happening:

Original:
pytest "tests/ttnn/unit_tests/operations/fused/test_layer_norm.py::test_large_layer_norm_with_weight_bias_and_residual_input[use_welford=False-w=4096-h=32]"
500 cycles a tile = 64k cycles
3 passes plus a write pass = ~4 passes = 256k cycles

Actual 
229168
/ 4 / 128 = 448
about 250 cycles a pass?

Welford's
pytest "tests/ttnn/unit_tests/operations/fused/test_layer_norm.py::test_large_layer_norm_with_weight_bias_and_residual_input[use_welford=True-w=4096-h=32]"

128 * (800+2*450) = 128 * 1700 = 217k cycles
actual
4524194 = 35k per tile

On branches:
rmillerTT/26077_welford_layernorm_infra commit 540e5063e7
vsureshTT/Welfords_LLK commit dc4ecb5d

Note that I had to remove the skip from the tests

**Comment by amahmudTT** [comment]
Ok I am more or less done with what i was working on, i will be looking at this starting tomorrow.

**Comment by bbradelTT** [comment]
One more strange behaviour:

Branches:
rmillerTT/26077_welford_layernorm_infra on tt-metal
vsureshTT/Welfords_LLK on tt-llk
as of Aug 22 2025

You can reproduce the short init issue by doing the following:
In the test ttnn/unit_tests/operations/fused/test_layer_norm.py::test_large_layer_norm_with_weight_bias_and_residual_input, call comp_pcc above the assert to print out the pcc and run the test. You'll see that the PCC is >0.9999 for all 4 h,w combinations
Go to the compute kernel layernorm_large_tensor_welford.cpp and find the transpose_wh_init() call. I left the short init call commented out along with a comment as to why we're using the full init. Comment out the full init and uncomment the short init
Re-run the pytest. I'm seeing PCCs of {0.97, 0.02, 0.99, 0.98}. A second run gives {0.98, 0.97, 0.99, 0.98}

**Comment by amahmudTT** [comment]
So to summarize, there are three issues with the relevant kernel. 

1. Loss of pcc when using short init for transpose
2. Loss of pcc when using short init for cast
3. Performance issue. 

For now will tackle the issues in order only inside this issue. If it is decided somebody else will look into this issue, then we can break the issues at that point.

**Comment by amahmudTT** [comment]
So in summary there are 3 errors with the new algorithm. 

1. ND PCC drop when using short transpose init on wormhole https://github.com/tenstorrent/tt-metal/issues/27334
2. PCC drop when using short/long broadcast init function
3. This weird performance behavior. 

After talking with Suresh and Borys, it seems issue nubmer 1 has the highest priority , I am moving to that issue and will come back to it sequentially.

**Comment by bbradelTT** [comment]
Update:
- The PCC issues seem to be addressed. We're still evaluating one last scenario.
- Most of the weird performance behaviour seems to have gone away once replay was used. The rest it seems to be due to the division.
- We encountered a new ND hang when using the LLK in GroupNorm. @vsureshTT @nsorabaTT please post clear repro instructions including branches, pytests to run, and exhibited behaviour.


The PCC issues have not been addressed, it has a deterministic component that is easily visible that one is addressed, the non deterministic part is not.

**Comment by nsorabaTT** [comment]
1. Clone this branch: https://github.com/tenstorrent/tt-metal/pull/27439 (nsorabaTT/gn_welford)
2. After forking this branch, also change the LLK branch
`cd tt_metal/third_party/tt_llk`
`git checkout vsureshTT/Welfords_LLK`
3. Run the following command: `timeout 1m pytest tests/ttnn/unit_tests/operations/fused/test_group_norm_DRAM.py::test_group_norm_DRAM -x`
The first case in this test is where a single core normalizes a single group. The group has 32 channels and 512*32 elements.
4. Navigate to `ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/compute/welford_groupnorm.cpp` and comment/uncomment out `welford` call in like 346 to see the impact of adding or removing the call.

**Comment by nsorabaTT** [comment]
Update:
1. Removing a spurious wait front in the compute kernel is making the hang go away. Will add another note once I run more tests to confirm.

**Comment by nsorabaTT** [comment]
The hang only occurs when there are multiple groups to be processed by each core. @vsureshTT is investigating if there are changes needed to address this.

**Comment by amahmudTT** [comment]
Okay the performance issue has been found. Instead of simulating which would give me very hard to interpret answer I surrounded the float division by 100s of nops before and after, the compiler probably rearranged some code but we can easily spot a call to a function to emulate hardware division and there are lots of other peripheral calls to facilitate that. Tensix cores do not support native hardware division. 

```
    asm volatile("nop");
    624c:       00000013                nop
    asm volatile("nop");
    6250:       00000013                nop
    asm volatile("nop");
    6254:       00000013                nop
    asm volatile("nop");
    6258:       00000013                nop
    asm volatile("nop");
    625c:       00000013                nop
    asm volatile("nop");
    6260:       00000013                nop
    asm volatile("nop");
    6264:       00000013                nop
    asm volatile("nop");
    6268:       00000013                nop
    asm volatile("nop");
    626c:       00000013                nop
    asm volatile("nop");
    6270:       00000013                nop


--------------------------------------   THIS SEGMENT SEEMS TO BE DOING THE DIVISION ---------------------------------

    float inv_n_plus_1 = 1.0/((float)N+1.0);
    6274:       000b8513                mv      a0,s7
    6278:       7c5010ef                jal     ra,823c <__floatunsisf>

0000627c <.LVL126>:
    627c:       0a8020ef                jal     ra,8324 <__extendsfdf2>

00006280 <.LVL127>:
    6280:       00040613                mv      a2,s0
    6284:       00048693                mv      a3,s1
    6288:       0f8010ef                jal     ra,7380 <__adddf3>

0000628c <.LVL128>:
    628c:       00050613                mv      a2,a0
    6290:       00058693                mv      a3,a1
    6294:       00040513                mv      a0,s0
    6298:       00048593                mv      a1,s1
    629c:       0d1010ef                jal     ra,7b6c <__divdf3>                   ------- > **SOFTWARE CALL**

000062a0 <.LVL129>:
    62a0:       160020ef                jal     ra,8400 <__truncdfsf2>


--------------------------------------   XXXXXXXXXXXXXXXXXXXXXXX ---------------------------------



000062a4 <.LVL130>:
    uint16_t high16_bits = static_cast<uint16_t>(bits >> 16);
    62a4:       01055693                srli    a3,a0,0x10
    TT_SFPLOADI(p_sfpu::LREG7, 8, high16_bits);
    62a8:       014686b3                add     a3,a3,s4
    62ac:       00dd2023                sw      a3,0(s10)
    62b0:       14281182                sfpadd  L6,L10,L0,L4,0
    TTI_SFPADD(p_sfpu::LCONST_1/*LREG10 = <1>*/, InputLREG, p_sfpu::LREG4, p_sfpu::LREG6,0);
    62b4:       d2fe0101                sfpmuli L4,-16512,0
    TT_SFPLOADI(p_sfpu::LREG7, 10, low16_bits);
    62b8:       00a97533                and     a0,s2,a0

000062bc <.LVL131>:
    62bc:       01350533                add     a0,a0,s3
    62c0:       00ad2023                sw      a0,0(s10)
    62c4:       1019d182                sfpmad  L6,L6,L7,L4,0
    TTI_SFPMAD(p_sfpu::LREG6, p_sfpu::LREG7, p_sfpu::LREG4, p_sfpu::LREG6, 0);
    62c8:       102d0102                sfpmad  L4,L11,L4,L0,0
    TTI_SFPMAD(p_sfpu::LREG11, p_sfpu::LREG4, InputLREG, p_sfpu::LREG4, 0);
    62cc:       102d8002                sfpmad  L0,L11,L6,L0,0
    TTI_SFPMAD(p_sfpu::LREG11/*LREG11 = <-1>*/,p_sfpu::LREG6, InputLREG, InputLREG,0);
    62d0:       d2fe0141                sfpmuli L5,-16512,0
    TTI_SFPMULI(negative_one_BF16, p_sfpu::LREG5, 0);
    62d4:       3c000002                sfpnop
    TTI_SFPNOP;
    62d8:       10101402                sfpmad  L0,L4,L0,L5,0
    TTI_SFPMAD(p_sfpu::LREG4, InputLREG, p_sfpu::LREG5, InputLREG, 0);
    62dc:       d2fe0141                sfpmuli L5,-16512,0
    TTI_SFPMULI(negative_one_BF16, p_sfpu::LREG5, 0);
    62e0:       3c000002                sfpnop
    TTI_SFPNOP;
    62e4:       1001d542                sfpmad  L5,L0,L7,L5,0
    TTI_SFPMAD(InputLREG, p_sfpu::LREG7, p_sfpu::LREG5, p_sfpu::LREG5, 0);
    62e8:       1829a502                sfpmul  L4,L10,L6,L9,0
    TTI_SFPMUL(p_sfpu::LCONST_1 /*LREG11 = <-1>*/, p_sfpu::LREG6, p_sfpu::LCONST_0, p_sfpu::LREG4, 0);
    62ec:       3c000002                sfpnop

000062f0 <.LBB2813>:
    TTI_SFPNOP;
    62f0:       d2fe0101                sfpmuli L4,-16512,0
    asm volatile("nop");
    62f4:       00000013                nop
    asm volatile("nop");
    62f8:       00000013                nop
    asm volatile("nop");
    62fc:       00000013                nop
    asm volatile("nop");
    6300:       00000013                nop
    asm volatile("nop");
    6304:       00000013                nop
    asm volatile("nop");
    6308:       00000013                nop
    asm volatile("nop");
    630c:       00000013                nop
    asm volatile("nop");
    6310:       00000013                nop
    asm volatile("nop");
    6314:       00000013                nop
```


If you want to dig into the function you can download and install sfpi compiler, then go to file. 

```./sfpi/gcc/libgcc/soft-fp/divdf3.c``` the function is written using macros and the macros are in ```./sfpi/gcc/libgcc/soft-fp/soft-fp.h``` . You can also look at the functions which are used to support the division in the assembly and find their implementation in the sfpi folder. The files for all the relevant functions are

```
00000000 l    df *ABS*  00000000 adddf3.c
00000000 l    df *ABS*  00000000 divdf3.c
00000000 l    df *ABS*  00000000 floatunsisf.c
00000000 l    df *ABS*  00000000 extendsfdf2.c
00000000 l    df *ABS*  00000000 truncdfsf2.c
```

Since it is a long sequence of riscv instructions 300-400 cycles is expected. @cmaryanTT
